message/validation/validator.go

```diff
@@ -0,0 +1,467 @@
+
+import (
+	"bytes"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	specqbft "github.com/bloxapp/ssv-spec/qbft"
+	spectypes "github.com/bloxapp/ssv-spec/types"
+	"golang.org/x/exp/slices"
+
+	"github.com/bloxapp/ssv/networkconfig"
+	ssvmessage "github.com/bloxapp/ssv/protocol/v2/message"
+	"github.com/bloxapp/ssv/protocol/v2/qbft/roundtimer"
+	"github.com/bloxapp/ssv/protocol/v2/ssv/queue"
+	ssvtypes "github.com/bloxapp/ssv/protocol/v2/types"
+	registrystorage "github.com/bloxapp/ssv/registry/storage"
+)
+
+const (
+	// lateMessageMargin is the duration past a message's TTL in which it is still considered valid.
+	lateMessageMargin = time.Second * 3
+
+	// clockErrorTolerance is the maximum amount of clock error we expect to see between nodes.
+	clockErrorTolerance = time.Millisecond * 50
+
+	maxMessageSize             = 8388608        // TODO: define
+	maxConsensusMsgSize        = maxMessageSize // TODO: define
+	maxPartialSignatureMsgSize = maxMessageSize // TODO: define
+	allowedRoundsInFuture      = 2
+	allowedRoundsInPast        = 1
+)
+
+var (
+	ErrUnknownValidator = errors.New("unknown validator")
+	ErrInvalidRole      = errors.New("invalid role")
+	ErrEarlyMessage     = errors.New("early message")
+	ErrLateMessage      = errors.New("late message")
+	ErrNoSigners        = errors.New("no signers")
+)
+
+type ConsensusID struct {
+	PubKey phase0.BLSPubKey
+	Role   spectypes.BeaconRole
+}
+
+type ConsensusState struct {
+	Signers map[spectypes.OperatorID]*SignerState
+}
+
+type MessageValidator struct {
+	netCfg       networkconfig.NetworkConfig
+	index        map[ConsensusID]*ConsensusState
+	shareStorage registrystorage.Shares
+}
+
+func NewMessageValidator(netCfg networkconfig.NetworkConfig, shareStorage registrystorage.Shares) *MessageValidator {
+	return &MessageValidator{
+		netCfg:       netCfg,
+		index:        make(map[ConsensusID]*ConsensusState),
+		shareStorage: shareStorage,
+	}
+}
+
+func (mv *MessageValidator) ValidateMessage(ssvMessage *spectypes.SSVMessage) (*queue.DecodedSSVMessage, error) {
+	if len(ssvMessage.Data) == 0 {
+		return nil, fmt.Errorf("empty data") // ERR_NO_DATA
+	}
+
+	if len(ssvMessage.Data) > maxMessageSize {
+		return nil, fmt.Errorf("message is too big, got %d bytes, limit %d bytes", len(ssvMessage.Data), maxMessageSize)
+	}
+
+	if !bytes.Equal(ssvMessage.MsgID.GetDomain(), mv.netCfg.Domain[:]) {
+		return nil, fmt.Errorf("wrong domain, got %v, expected %v", hex.EncodeToString(ssvMessage.MsgID.GetDomain()), hex.EncodeToString(mv.netCfg.Domain[:]))
+	}
+
+	if !mv.validRole(ssvMessage.MsgID.GetRoleType()) {
+		return nil, ErrInvalidRole
+	}
+
+	share := mv.shareStorage.Get(ssvMessage.MsgID.GetPubKey())
+	if share != nil {
+		return nil, ErrUnknownValidator // ERR_VALIDATOR_ID_MISMATCH
+	}
+
+	if !share.BeaconMetadata.IsAttesting() {
+		return nil, fmt.Errorf("validator is not active")
+	}
```

Comment:
Does this encompass the case for a liquidated validator? I mean, if the validator is liquidated, will this raise the error?

Reply:
yes, `IsAttesting` means it's either `ValidatorStateActiveOngoing` or `ValidatorStateActiveExiting`